### ISSUE TITLE
Autologin option not available in KDE Wayland

### ISSUE THREAD

**Issue by ganomi46** [keepassxreboot/keepassxc#2281]
Autologin option not available in KDE Wayland

## Expected Behavior
Autologin option should be available as under X.

## Current Behavior
Autologin option not available in KDE Wayland

## Steps to Reproduce (for bugs)
Login to Plasma Wayland
Open KeePassXC
You will there are no auto login options

Operating system: OS
Debian SID amd64

**Comment by droidmonkey** [comment]
Are you referring to AutoType??

**Comment by ganomi46** [comment]
Hi.Yes, thatbis whatnI meant.-JiriOn 12 Sep. 2018 2:14 am, Jonathan White <notifications@github.com> wrote:Are you referring to AutoType??

**Comment by Whompithian** [comment]
I can confirm lack of Auto-Type options using Plasma Wayland.

OS: 4.14.65-gentoo x86_64
plasma-5.46.0
wayland-1.15.0

**Comment by phoerious** [comment]
Auto-Type cannot work on Wayland due to security restrictions for which we haven't found a solution yet.

**Comment by rockihack** [comment]
I implemented autotype for wayland, still needs some work though.

https://github.com/rockihack/keepassx/blob/wayland-autotype/wayland.txt
https://github.com/rockihack/keepassx/tree/wayland-autotype/src/autotype/wayland

**Comment by apollo13** [comment]
It might be worth to look what https://gitlab.gnome.org/ofourdan/gnome-ponytail-daemon does. I seems to allow writing arbritrary key sequences to windows.

**Comment by tmccombs** [comment]
fwiw, there is a keepass plugin that supports this (https://keepass.info/help/kb/autotype_wayland.html), it does require access to /dev/uinput though. 

For wlroots based compositors (such as sway) the input-method protocol (https://github.com/swaywm/wlroots/blob/master/protocol/input-method-unstable-v2.xml)  may allow implementing AutoType

**Comment by curiosityseeker** [comment]
I'm using KeePassXC on Arch Linux with Wayland (on KDE) in combination with [ydotool](https://github.com/ReimuNotMoe/ydotool), specifically [ydotool-git](https://aur.archlinux.org/packages/ydotool-git) in the AUR. And voilà - Autotype works flawlessly! 

So perhaps it's possible to integrate ydotool's approach in KeePassXC directly?

**Comment by droidmonkey** [comment]
Oh nice!!

**Comment by apollo13** [comment]
Mhm, seems like it has a client/server structure because it uses `/dev/uinput` which requires (usually) root permission. Still better than no autotype I guess :D

**Comment by curiosityseeker** [comment]
Yes, it seems to be similar how KeePass [does](https://keepass.info/help/kb/autotype_wayland.html) it with KPUInput.

**Comment by apollo13** [comment]
https://gitlab.com/dogtail/dogtail/ seems to use the accessibility features in gnome and kde to send text to applications without the need of uinput and as such without any extra daemon.

**Comment by apollo13** [comment]
Ok, dogtail uses ponytail on wayland which I linked above already. Seems to be a rather heavy solution :/

**Comment by droidmonkey** [comment]
I think I'll opt for https://github.com/keepassxreboot/keepassxc/issues/2281#issuecomment-425932745 solution

**Comment by phoerious** [comment]
Using the DE's accessibility features is the only correct way to implement it. I only wish there were some kind of standard interface and not several largely incompatible ones, like it always is on Linux.

**Comment by MAFLO321** [comment]
I'm not very familiar with Wayland in deep, but can [text-input-v3](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/blob/master/unstable/text-input/text-input-unstable-v3.xml) protocol be used to solve the virtual-input problem?
KDE Plasma's and Gnome's Compositors both support it.

already mentioned by https://github.com/keepassxreboot/keepassxc/issues/2281#issuecomment-657179359

**Comment by phoerious** [comment]
Perhaps, need to check. Thanks for the pointer.

**Comment by ztNFny** [comment]
> I'm using KeePassXC on Arch Linux with Wayland (on KDE) in combination with [ydotool](https://github.com/ReimuNotMoe/ydotool), specifically [ydotool-git](https://aur.archlinux.org/packages/ydotool-git) in the AUR. And voilà - Autotype works flawlessly!
> 
> So perhaps it's possible to integrate ydotool's approach in KeePassXC directly?

Could you elaborate on that? How can KeePassXC be configured to use ydotool? I thought to just use it with cmd://, but then the input just goes .. no idea where.

**Comment by sam9032** [comment]
I don't have KDE Plasma to test right now, but under Gnome 40 with wayland it works ootb. Is this issue still present in plasma?

**Comment by vimpostor** [comment]
> I don't have KDE Plasma to test right now, but under Gnome 40 with wayland it works ootb. Is this issue still present in plasma?

I just tested in Plasma and it doesn't work there. Neither global autotype nor manually choosing autotype from Keepass itself are working.
I am curious, does anyone know how this is working in Gnome? As far as I can see in the codebase, there is no implementation merged for Wayland yet, so I don't get why this is working in Gnome ootb for you.

Tested on Plasma 5.21.4 on Arch btw.

**Comment by curiosityseeker** [comment]
> I just tested in Plasma and it doesn't work there. Neither global autotype nor manually choosing autotype from Keepass itself are working.

> Tested on Plasma 5.21.4 on Arch btw.

I'm also using Plasma with Wayland on Arch, and KeepassXC works flawlessly in combination with the KeePassXC-Browser add-on both in Firefox and Brave.
If it doesn't for you: are you using Firejail?

@ztNFny : I removed ydotool - it seems it's not (no longer?) necessary.

**Comment by ztNFny** [comment]
> I'm also using Plasma with Wayland on Arch, and KeepassXC works flawlessly in combination with the KeePassXC-Browser add-on both in Firefox and Brave.

This issue is about AutoType, not browser extensions. That's a separate mechanism using native msgs.

**Comment by sam9032** [comment]
I have now tested the whole thing with Gnome 40 again in more detail. So under X-org as well as under Wayland.

I tested the global autotype function under the "General" tab and the entry based autotype function both with different websites and the Spotify client.
I leave out the browser addon, as it is not part of this issue.

As far as I can tell for my test, both (x-org and wayland) work almost the same.
No matter how I have my login field filled in (global or entry), both manage to type in the selected entry.
Under Wayland, both have a problem with the "@" symbol in the username/password for certain applications (ex: Spotify) and just don't type that.
In Firefox, having web pages type something in is not a problem.

Under both Wayland and X-Org the global function only shows entries with matching title, links do not work.
Typing the selected ones then works again without problems as long as the fokus is on the first input field (username).

**Comment by droidmonkey** [comment]
Thank you for all the input but we don't need any further discussion on this. What you are seeing is Auto-Type happening between applications running on xwayland backend. You are still communicating through an X11 server between keepassxc and the other application. Wayland is just receiving the output and passing it to the compositor.

**Comment by CIAvash** [comment]
I know it's been said that there is no need for further discussion, but if keepassxc knows the window is running on wayland or xwayland or X11, then it can do something like using [wtype](https://github.com/atx/wtype) for wayland windows and `xdotool type` for X11.

I tried and it works, you just have to know the type of window.

**Comment by zsolt-donca** [comment]
> ... if keepassxc knows the window is running on wayland or xwayland or X11, then it can do something like using [wtype](https://github.com/atx/wtype) for wayland windows and `xdotool type` for X11.

The [ydotool](https://github.com/ReimuNotMoe/ydotool) also works in a display server-agnostic way, working both on Xorg and on Wayland.

I haven't tried this personally, though.

**Comment by droidmonkey** [comment]
> This program requires access to /dev/uinput. This usually requires root permissions.

**Comment by Aetf** [comment]
Maybe it's possible to implement the input method protocol under Wayland as a way to perform autotype?

https://dcz_self.gitlab.io/posts/input_method/#:~:text=The%20purpose%20of%20an%20input,and%20gives%20it%20to%20applications

**Comment by Zocker1999NET** [comment]
I use sway and use  https://github.com/atx/wtype (uses input method protocol of Wayland, as I know) to insert keys, works without root and really easy. Maybe this could be used to implement it here?

**Comment by MarkJaroski** [comment]
I understand from droidmonkey's post that additional commentary isn't necessarily welcome on this issue, however this is the only thing right now preventing me from moving to Wayland, so I'd like to at least have some understanding of what the blockers are, and whether they're more of a global problem with Wayland, or if they're just a KeepassXC problem. Again, I apologise in advance if I'm creating noise, but it would be good to have a little bit of information about what if anything could move this forward.

**Comment by droidmonkey** [comment]
What can move this forward?

A wayland expert that programs a virtual keyboard interface for keepassxc and some way to pull window titles from the desktop environment that works across all variations of Linux desktop envs (gnome mutter, kde kwin, Sway, etc etc).

Wayland has taken the stance of "Wayland is not a compositor" and has ceded that responsibility to the various desktop environments. This creates the situation where each desktop env has its own set of extended protocols and internal interface api (and maybe even permissions) to poll for information that was openly available via X11 server in the past. This includes window titles and switching between windows. Wayland has taken the stance that every client is isolated and should have no knowledge of any other client. This means keepassxc cannot "find out" about other windows unless the compositor exposes that information in some other fashion. 

So this isn't really a "why can't you do this in wayland" discussion because that is not where you do this anyway. Wayland is just a protocol for drawing windows and sending/receiving inputs to the compositor. 

https://way-cooler.org/book/wayland_introduction.html

https://stackoverflow.com/a/64030239

**Comment by MarkJaroski** [comment]
Would it help, maybe, to get freedesktop.org's attention?

**Comment by droidmonkey** [comment]
No it wouldn't help in any way, they are not going to change the protocol to violate its core tenants. Further, each compositor can choose not to implement portions of the protocol that may expose client to client information. GNOME has already taken this stance.

**Comment by MarkJaroski** [comment]
So, I guess there is no generic way forward?

What about doing something completely different, something like a browser plugin, but at the desktop level?

I realize that I'm getting off-topic now, but if you have any ideas I might be able to help hack some kind of POC together.

**Comment by MarkJaroski** [comment]
Here's what might be a dumb idea: what about using the virtual keyboard interface? Keepass2Android does this, and while it's a bit of a pain in the neck to switch between keyboards every time you want to type a secret it works.

**Comment by ccoenen** [comment]
[Barrier](https://github.com/debauchee/barrier/issues/109) and [Synergy](https://github.com/symless/synergy-core/issues/4090#issuecomment-831825040) have the same problem and it seems they also don't have a clear way forward right now.

**Comment by tmccombs** [comment]
I think there is a path forward for wlroots-based DEs and possibly KDE. But it seems unlikely GNOME will let non-GNOME apps do this sort of thing anytime soon.

**Comment by MarkJaroski** [comment]
Here's a left field idea: what about virtual USB?

**Comment by foobar13372** [comment]
Are there any news here? Working approaches/workarounds?

**Comment by slagiewka** [comment]
Has anyone considered [libei](https://gitlab.freedesktop.org/libinput/libei)?

**Comment by droidmonkey** [comment]
Keyboard input is actually not the challenge here. The issue is window title harvesting and ability to read the window title to use global auto type capability. That is currently _impossible_ through the Wayland protocol and no Wayland based compositors have exposed this as a standard or defined interface.

**Comment by lennartfricke** [comment]
But wouldn't it be ok, if the auto-type feature is limited to typing into the last focused text-field/-area. It would mean one has to select the correct entry beforehand or in a selection frame displayed on global auto-type. But at least some kind of auto-type would work. Or am I missing something here?

**Comment by apollo13** [comment]
I think that would at least be a middleground till the rest of the missing autotype functionality is possible (ie querying the window list and active windows)

**Comment by Ferk** [comment]
An alternative would be to add an option to `keepassxc` commandline (or `keepassxc-cli`) in which you can pass the title of the window as parameter and have it open the autotype dialog.

Then rely on the compositor/user to configure a shortcut to call the command appropriatelly with the required info.

At least in Sway it would be viable to configure it that way (and in fact, this might be the only way to set a "global shortcut" since I'm not sure KeepassXC is able to configure such shortcut from its settings, it's not working right now in Sway). And I expect the same from any compositor that allows any sort of flexibility / scripting for its shortcuts. If they don't offer it then it'd be their fault, imho.

Maybe for Gnome a Gnome Shell extension could be written that calls the same autotype command (I expect Gnome Shell should be able to support that at least).

**Comment by Aetf** [comment]
> An alternative would be to add an option to `keepassxc` commandline (or `keepassxc-cli`) in which you can pass the title of the window as parameter and have it open the autotype dialog.

Or as a dbus api. It will be usable in kde via kwin script, then. And you can still call it on command line using dbus-send or busctl

**Comment by MarkJaroski** [comment]
I've bitten the bullet to speak, and have started developing integrations everywhere using this, with the goal of making autotype unnecessary:

https://github.com/hargoniX/keepassxc-proxy-client

**Comment by droidmonkey** [comment]
Sweet! Good solution

**Comment by Ferk** [comment]
> https://github.com/hargoniX/keepassxc-proxy-client

As far as I understand, it only supports looking up entries based on their URLs,  not the Window Titles configured in the auto-type associations, which might also have different key sequences for different Windows (or more than one key sequence, like for windows requesting the TOTP).

Is it possible to query entries based on the Window Titles via keepassxc-proxy? what about obtaining the configured key sequences?

**Comment by MarkJaroski** [comment]
Well, it's not autotype, so it really isn't about the window, rather the idea is to find a hook in the application itself, or in the worst case to set up a global key binding.

**Comment by rldleblanc** [comment]
As a work around for KDE on Wayland (Debian Bookworm) is to run KeepassXC through Xwayland. I was able to do this with `WAYLAND_DISPLAY=no keepassxc` and it seems to have restored global AutoType shortcuts. I've only tried it on a couple of windows though. Thought I'd share.

**Comment by rldleblanc** [comment]
As a follow-up, `GDK_BACKEND=x11` did not seem to work and to update the Application Menu item, I had to enter `env WAYLAND_DISPLAY=no keepassxc %f` in the `Command` section. Hopefully that will help buy some time and still be able to test/use Wayland.

**Comment by rusty-snake** [comment]
> GDK_BACKEND=x11 did not seem to work

Because kpxc is a Qt app and not a GTK one. `QT_QPA_PLATFORM=xcb` is the right environment variable. See https://wiki.archlinux.org/title/Wayland#Qt.

> As a work around for KDE on Wayland (Debian Bookworm) is to run KeepassXC through Xwayland.

As long as you only need to auto-type into XWayland clients yes but if you need to auto-type into native Wayland clients ...

**Comment by rldleblanc** [comment]
You are correct. I thought I read that Google Chrome was defaulting to Wayland (which it is not) and so when it worked I thought there was something there. Starting Chrome with Wayland did not allow AutoType to work (neither did Konsole work). I thought after a few years I would give Wayland a try, it didn't crash immediately and Dvorak now worked, but I think I will try it again in a few years.

**Comment by dimztimz** [comment]
> Keyboard input is actually not the challenge here. The issue is window title harvesting and ability to read the window title to use global auto type capability. That is currently _impossible_ through the Wayland protocol and no Wayland based compositors have exposed this as a standard or defined interface.

Maybe you can first try to implement auto-type of manually selected entry (right click -> Perfom Auto-Type). Is that a problem with Wayland because it requires focusing the previously focused window?

**Comment by droidmonkey** [comment]
Yes it is for that very reason. Although we can just minimize our own window and hope the focus shifts correctly.

**Comment by dimztimz** [comment]
Just to recap, there are 3 problems related to this issue.

1. Keyboard input. There are multiple approaches for solving this. One is by using `/dev/uinput`. See how KeePass does it https://keepass.info/%0D/help/kb/autotype_wayland.html or a comment above or some [stackoverflow](https://stackoverflow.com/questions/51406276/simulating-mouse-and-keyboard-input-on-wayland-and-x11). Does Qt has something for this? Maybe use the clipboard instead of simulating keyboard input? Libei?
2. Focusing the last focused window. What is used now? What does KeePass do? Does it minimizes the window?
3. Reading the titles of the windows. There is no straightforward solution for this.

Feature 1 and 2 are needed for auto-type of selected entry, 1 and 3 are needed for global auto-type. Did I missed something?

**Comment by droidmonkey** [comment]
Registering a global shortcut, but that might be covered by existing Wayland api. Most likely not, which would make it desktop env specific. /dev/uinput is not an option, requires root.

**Comment by rusty-snake** [comment]
> Registering a global shortcut, but that might be covered by existing Wayland api. Most likely not, which would make it desktop env specific.

FWIW https://github.com/flatpak/xdg-desktop-portal/issues/624

**Comment by ibahnasy** [comment]
This is also affected in Gnome (Ubuntu 22.10).

**Comment by cskr** [comment]
Auto-type from context menu used to work fine when running Firefox under xwayland, but even this has stopped working since 2.7.0.

**Comment by dasistwas** [comment]
Ubuntu 22.04 Beta, Keepassxc 2.7.0:

- Auto-Type window selection working for: Signal, Chromium,
- Not working for: Firefox, Terminal, Thunderbird

**Comment by tmccombs** [comment]
@dasistwas that is likely because Signal and Chromium are using XWayland, but Firefox, Terminal, and Thunderbird are using native wayland.

**Comment by dasistwas** [comment]
Update from Ubuntu 22.04 Beta: Suddenly auto-type started to work. But version of keepassxc did not change. It is still 2.7.1

**Comment by ourichermath** [comment]
I'd like to request the documentation at:

https://keepassxc.org/docs/KeePassXC_UserGuide.html#_auto_type

be added to.  I'm on Fedora (Wayland) and it was really difficult to figure out where and how to change either the command line (-platform xcb) for my KeepassXC that was set to Autorun or the environment variable (QT_QPA_PLATFORM=xcb).  Some mention of where and how would be nice.

Also, the documentation isn't clear in what KPXC is looking for with AutoType to work with applications.  For instance, when I try to set it up to work with the password entry in the Virtual Machine Manager or for Grsync, the window names don't show up in KPXC's Advanced > Windows dropdown.  And, no matter what I manually type in for a Window name, it doesn't work.

**Comment by foobar13372** [comment]
Ubuntu 22.04 + snap version of Firefox (Ubuntu 22.04 upgrademigrates FF to the snap version automatically): autotype works again. Did not test other programs since I only need keepass for Firefox.

**Comment by ibahnasy** [comment]
Wayland is disabled by default in 22.04 because of a bug, may be that's why autotype is working?

------- Original Message -------

**Comment by dasistwas** [comment]
> Wayland is disabled by default in 22.04 because of a bug, may be that's why autotype is working?

no, I am on Wayland. But just realized seems only Firefox and xwayland programs do work with autotype. Terminal does not for example.

**Comment by tmccombs** [comment]
> only Firefox and xwayland programs
is firefox using xwayland?

**Comment by dinotheextinct** [comment]
Is there any news to this? I am on sway and would like to know how I can get autotype working for firefox. Setting the xcb plattform, enables autotype, but hitting the configured shortcut from keepassxc for autotype doesnt open anything.

**Comment by Zocker1999NET** [comment]
> but hitting the configured shortcut from keepassxc for autotype doesnt open anything.

As I know it, Sway does not allow other applications to apply global shortcuts themselves and also I do not know of any Wayland protocol which would implement this. It might be required to configure the shortcut on Sway manually and trigger AutoType on KeePassXC using a command / D-Bus (which is not possible yet?)

**Comment by siccovansas** [comment]
I recently installed Ubuntu 22.04 and ran into this issue.

Firefox and Chromium apparently use XWayland, so KeePasXC's Global Auto-Type works there.

My main other use of Auto-Type is in the terminal for which I use the [Terminator](https://gnome-terminator.org/) terminal. I use the following (hopefully temporary) workarounds to run Terminator using XWayland when using the Ctrl + Alt + T shortcut and when starting Terminator via its icon in the launcher:

- In Ubuntu's Settings > Keyboard > Keyboard Shortcuts add this Custom Shortcut:
  - Name: Terminator with XWayland (to support KeePassXC)
  - Command: sh -c 'WAYLAND_DISPLAY=no /usr/bin/terminator'
  - Shortcut: Ctrl + Alt + T
- Create the file `~/bin/terminator` with this code
  - ```
    #!/bin/sh

    # Catches calls to run 'terminator' (as this executable is in ~/bin and thus
    # listed earlier in PATH) and run it with XWayland which makes KeePassXC work
    # (via https://askubuntu.com/a/923393/54352)
    export WAYLAND_DISPLAY=no
    exec /usr/bin/terminator "$@"
    ```
  - Make it executable: `chmod +x ~/bin/terminator`

**Comment by dinotheextinct** [comment]
Not sure if this is some Gnome garbo, but under KDE Wayland, KeepassXC doesnt even offer the Autotype option??

**Comment by rusty-snake** [comment]
https://github.com/keepassxreboot/keepassxc/blob/861fe2e5a9ec1025f89591b14017fc3880fb3733/docs/topics/AutoType.adoc?plain=1#L10

**Comment by dinotheextinct** [comment]
So I got the autotype option back, but when hitting the shortcut, there is no popup on wayland, when using it for firefox...

**Comment by ibahnasy** [comment]
I can confirm that auto typing is currently working in Wayland normally after the past system updates released in the past few weeks.

**Comment by dasistwas** [comment]
> I can confirm that auto typing is currently working in Wayland normally after the past system updates released in the past few weeks.

Did you test autotype with the terminal as well? Only xwayland programs seem to work. All wayland windows do not trigger the autotype dialogue when using the global autotype shortcut.

**Comment by ibahnasy** [comment]
> > I can confirm that auto typing is currently working in Wayland normally after the past system updates released in the past few weeks.
> 
> Did you test autotype with the terminal as well? Only xwayland programs seem to work. All wayland windows do not trigger the autotype dialogue when using the global autotype shortcut.

It doesn't work with the terminal.
Triggering the autotype from the KPXC window or using the keyboard key combination, it works fine with some GUI apps like Google Chrome, but apps like FIrefox or Nautilus doesn't work.

**Comment by pio2398** [comment]
> > > I can confirm that auto typing is currently working in Wayland normally after the past system updates released in the past few weeks.
> > 
> > 
> > Did you test autotype with the terminal as well? Only xwayland programs seem to work. All wayland windows do not trigger the autotype dialogue when using the global autotype shortcut.
> 
> It doesn't work with the terminal. Triggering the autotype from the KPXC window or using the keyboard key combination, it works fine with some GUI apps like Google Chrome, but apps like FIrefox or Nautilus doesn't work.

It's just can't work with wayland app. Could you read at least few comment back before posting? https://github.com/keepassxreboot/keepassxc/issues/2281#issuecomment-831369450

**Comment by bigdiff** [comment]
Would it be possible to implement a workaround for now till xdg-desktop-portals implements a global shortcut portal.

I was thinking about making keepassxc launch with a --auto-type flag. Adding that to the list of custom shortcuts in gnome-control-panel would resolve the issue for now.

```
name: KeePassXC Auto-Type
command: keepassxc --auto-type
```

![gnome control center > add custom shortcuts](https://user-images.githubusercontent.com/96084563/178689496-6d15c430-d8f5-4a33-8820-feef7a22542c.png)

**Comment by vimpostor** [comment]
> Would it be possible to implement a workaround for now till xdg-desktop-portals implements a global shortcut portal.

Can you guys at least [read through the thread](https://github.com/keepassxreboot/keepassxc/issues/2281#issuecomment-988874218) before pushing this again. Even if you fix the shortcut registering, you still can't find out the title of the window.

[And again](https://github.com/keepassxreboot/keepassxc/issues/2281#issuecomment-922332932), the easiest way to move this issue forward is to move the already existing upstream proposals forward:
- https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/56
- https://github.com/flatpak/xdg-desktop-portal/issues/624

Why are people pushing this issue every week but not the real underlying issue in the upstream protocols? This is all so backwards.
Quite frankly, this issue should be closed as "Won't fix" or locked until this mess is fixed in Wayland.
[Originally back in 2013 Wayland was meant to cut out the "useless" middleman](https://youtu.be/GWQh_DmDLKQ), ironically now we end up with even more useless middlemen just to deal with basic stuff like negotiating shortcut registering or server side decorations.

**Comment by droidmonkey** [comment]
This information was provided by @galaxysnail

Currently, we have the global keyboard shortcut portal and [libei](https://gitlab.freedesktop.org/libinput/libei), which should be good enough for auto-type.

For enumerating windows, there is a [foreign toplevel list protocol](https://wayland.app/protocols/ext-foreign-toplevel-list-v1).

**Comment by droidmonkey** [comment]
Just learned that Goldwarden is using the libportal RemoteDesktop interface for auto-type. Unsure on the actual security of that, but here it is:

https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.RemoteDesktop.html

https://github.com/quexten/goldwarden/blob/main/cli/autotype/libportalautotype.go

**Comment by Thatoo** [comment]
On Gnome 46, I now have a new windows asking me if I want to authorized the operation but after syaing yes, nothing more happen. Normal?

**Comment by droidmonkey** [comment]
Gnome added that

**Comment by Thatoo** [comment]
It doesn't help to make AutoType working on Wayland yet?

**Comment by sagebind** [comment]
No. KeePassXC will have to implement _something_ new for AutoType to work on Wayland (even if what that is exactly is not clear). There's no way that KeePassXC's existing X11 code will "magically" work with Wayland someday.

**Comment by Zahrun** [comment]
I’m offering $100 for a fix to this issue at gitpay https://gitpay.me/#/task/763/auto-type-not-available-when-using-wayland
If this issue is important to you, please feel free to add any amount to the bounty.

**Comment by TheConfuZzledDude** [comment]
> Just learned that Goldwarden is using the libportal RemoteDesktop interface for auto-type. Unsure on the actual security of that, but here it is:
> 
> https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.RemoteDesktop.html
> 
> https://github.com/quexten/goldwarden/blob/main/cli/autotype/libportalautotype.go

I believe that using the RemoteDesktop portal interface is fine for that, it's really just for accessing emulated inputs. The `ConnectToEIS` would be used to get the fd for the compositor's EIS, and then libei can be used to actually send inputs.

See also https://libinput.pages.freedesktop.org/libei/api/index.html#sec-oeffis for liboeffis which handles the desktop portal access and just gets the libei fd.

As for window detection/focusing, `ext-foreign-toplevel-list-v1` isn't really implemented anywhere yet, and activation of those foreign windows isn't close.

**Comment by zsteinmetz** [comment]
> I’m offering $100 for a fix to this issue at gitpay https://gitpay.me/#/task/763/auto-type-not-available-when-using-wayland If this issue is important to you, please feel free to add any amount to the bounty.

Just added 50 bucks :money_with_wings:

**Comment by zsteinmetz** [comment]
@droidmonkey, could you please add the bounty tag and the current amount to this issue to raise more awareness? Thank you!

**Comment by droidmonkey** [comment]
We don't officially back bounties anymore, but I added the tag for sport.

**Comment by TheConfuZzledDude** [comment]
I am working at this currently, I'll have a WIP PR up soonish

**Comment by TheConfuZzledDude** [comment]
[The PR is up now](https://github.com/keepassxreboot/keepassxc/pull/10905), still needs work

**Comment by reshadp** [comment]
When I run KeePassXC from the terminal using the command `keepassxc -platform xcb` I have get the auto type options enabled again (I know this is supposed to work).
I am on KDE Plasma 6 running Wayland and am able to use the autotype function when in this mode to type into even what should be Wayland windows (Konsole, Kate, Kwrite, Firefox) after accepting a dialogue box which asks me to accept a remote control request.
![Screenshot_20240722_120525](https://github.com/user-attachments/assets/58fcde11-472f-471b-abdb-a28824bb4ab7)
The dialogue box only shows up once per Plasma session (which I think is fine).

This dialogue box is relatively new (maybe a month or so) auto-type would only work with X11 apps running in XWayland (I used to auto-type into Chromium) before I noticed this dialogue box.
Hope this workaround helps anyone while the PR to get this properly implemented.

**Comment by agowa** [comment]
The documentation already coveres this: https://keepassxc.org/docs/KeePassXC_UserGuide#_auto_type

> Auto-Type will be disabled when run with a Wayland compositor on Linux. To use Auto-Type in this environment, you must set QT_QPA_PLATFORM=xcb or start KeePassXC with the -platform xcb command-line flag.


After testing this right now I only see two minor issues remaining to be fixed regarding this ticket:
* KeePass automatically detecting that it is running in Wayland mode and that it needs to use the features provided by this flag without requiring user action to specify it on startup.
* The Auto-Type didn't wait for me to actually confirm the sharing and by the time I hit "Share" it already tried to type half of the credentials. And thereby the credentials it entered were incomplete. This isn't an issue for any following attempts though.

**Comment by vimpostor** [comment]
I feel there is a need to explain why suddenly the XWayland -> Wayland client direction is working without changes from the Keepass side and why the problem is still only half solved:

The "Remote control requested" dialog you are seeing is a result of [KDE implementing](https://invent.kde.org/plasma/xdg-desktop-portal-kde/-/issues/12) the [Input Capture](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.InputCapture.html#input-capture) and [Remote Desktop](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.RemoteDesktop.html) portal together with support for the [ConnectToEIS](https://invent.kde.org/plasma/xdg-desktop-portal-kde/-/merge_requests/279) method.

[libei](https://gitlab.freedesktop.org/libinput/libei) is a library for emulated input on Wayland and has support for a superset of the functionality that was available in the [XTest extension](https://www.x.org/releases/X11R7.7/doc/xextproto/xtest.html), which [KeepassXC uses to implement autotype on X11](https://github.com/keepassxreboot/keepassxc/blob/afe634d4f2a0f32e96ea720289262a4be90afc0f/src/autotype/xcb/AutoTypeXCB.cpp#L378).

libei can be used as a drop-in replacement for XTest in XWayland and the [README even illustrates this architecture](https://gitlab.freedesktop.org/libinput/libei/-/blob/a424458f035687f405e7ad7c0a56b9b85a58cafa/README.md#xwayland-and-xtest).
Note that the comment about the implementations being in PoC state is outdated, [full support was merged into XWayland a year ago](https://gitlab.freedesktop.org/xorg/xserver/-/merge_requests/975).

This means that the fully working path from XWayland to Wayland clients is as follows:
- KeepassXC running in XWayland sends a `XTestFakeKeyEvent()` event
- XWayland retreives the XTest event and [resends it as an libei event](https://gitlab.freedesktop.org/xorg/xserver/-/blob/aec722db36a3e0b588aa6a27d8f77d10b6c92b27/hw/xwayland/xwayland-xtest.c#L686)
- [The libei implementation in kwin retreives the event and emits the key event](https://invent.kde.org/plasma/kwin/-/blob/7d3f58d0caee9182604b643083787e613b31f110/src/plugins/eis/eiscontext.cpp#L302)
- The currently active Wayland client retreives the input event via the normal input stack

As mentioned at the beginning, there are however some caveats with this:
1. This only works if KeepassXC is running in XWayland mode, there is no native Wayland support yet (however there is a [PR being worked on](https://github.com/keepassxreboot/keepassxc/pull/10905)).
2. It does not support filtering based on Window titles. This means it is not possible to hit a global key and show entries based on the currently active window, as is possible with Xorg. Instead the entry has to be manually selected. This caveat will also still be true for the pending native Wayland PR, there simply is no mechanism on Wayland to do that (and many compositors even refuse to implement [ext-foreign-toplevel-list-v1](https://wayland.app/protocols/ext-foreign-toplevel-list-v1#ext_foreign_toplevel_handle_v1:event:title))
3. The dialog asking for input consent will silently swallow all sent input events until the dialog is accepted, so on the first time the target may receive only half or none of the intended keyboard events.

So while the situation is certainly better than before (support for basic autotype is working in XWayland and soon in Wayland), full support like on Xorg with autotype triggered on a hotkey and automatically filtering based on window title is still far from a reality.

**Comment by agowa** [comment]
@vimpostor is https://wayland.app/protocols/ext-foreign-toplevel-list-v1#ext_foreign_toplevel_handle_v1:event:title part of the accessibility API? Getting information of running windows sounds like something almost every accessibility tool would need to function properly. And doesn't Wayland at least also have support for screen readers? How do they access such information then? Doesn't Orca Screen Reader already support Wayland? So aren't there already APIs that could be used?

Or is this all still a big work in progress mess?

**Comment by MRDGH2821** [comment]
This seems to be useful:
https://espanso.org/docs/install/linux/#adding-the-required-capabilities

**Comment by Ironwally** [comment]
I found this issue after installing keepassxc via snap. Up to now I used keepassxc via appimage. Here I did not have any issues or needed workarounds. 

So maybe it would be interessting to know why this works with the appimage, but not with, I suppose, the rest?

**Comment by MRDGH2821** [comment]
> I found this issue after installing keepassxc via snap. Up to now I used keepassxc via appimage. Here I did not have any issues or needed workarounds.
> 
> So maybe it would be interessting to know why this works with the appimage, but not with, I suppose, the rest?

You mean to say that Auto type works in wayland session when run as app image?
Because I tried & there is no Autotype option.

Can you confirm if you were in Wayland/X11 session?

**Comment by dasistwas** [comment]
On Ubuntu 24.04 I used to start keepassxc with` QT_QPA_PLATFORM=xcb /snap/bin/keepassxc -platform xcb %f` That worked for a week and now AutoType is gone again. Snap version of keepassxc is 2.7.8, then I tried the AppImage, but here it is the same (2.7.9) - AutoType does not work.

Launching AppImage:
```
./KeePassXC-2.7.9-x86_64.AppImage 
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
QObject::startTimer: Timers cannot have negative intervals
qt.network.ssl: QSslSocket: cannot resolve EVP_PKEY_base_id
qt.network.ssl: QSslSocket: cannot resolve SSL_get_peer_certificate
qt.network.ssl: QSslSocket: cannot call unresolved function SSL_get_peer_certificate
```

Launching snap:
```

QT_QPA_PLATFORM=xcb /snap/bin/keepassxc -platform xcb %f
Qt: Session management error: Could not open network socket
MESA: warning: Driver does not support the 0x7d55 PCI ID.
libGL error: glx: failed to create dri3 screen
libGL error: failed to load driver: iris

```

**Comment by Ironwally** [comment]
> > I found this issue after installing keepassxc via snap. Up to now I used keepassxc via appimage. Here I did not have any issues or needed workarounds.
> > So maybe it would be interessting to know why this works with the appimage, but not with, I suppose, the rest?
> 
> You mean to say that Auto type works in wayland session when run as app image? Because I tried & there is no Autotype option.
> 
> Can you confirm if you were in Wayland/X11 session?

Yes, I am using wayland.
I never specified any options like -platform or something. I just downloaded the app and started it.  Out of the box it worked fine.

I'll check if theres some argument run automatically when starting via appimage.

**Comment by Ironwally** [comment]
> Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.

Interesting, your startup errors for the appimage.

Mine is slightly different and makes me think that mine is actually running with xcb.
`qt.qpa.plugin: Could not find the Qt platform plugin "wayland" in ""` 
But I don't know how to interprete this one.
It seems to want to load wayland, but doesn't give the same error as yours.

**Comment by ioogithub** [comment]
> When I run KeePassXC from the terminal using the command `keepassxc -platform xcb` I have get the auto type options enabled again (I know this is supposed to work). I am on KDE Plasma 6 running Wayland and am able to use the autotype function when in this mode to type into even what should be Wayland windows (Konsole, Kate, Kwrite, Firefox) after accepting a dialogue box which asks me to accept a remote control request. ![Screenshot_20240722_120525](https://private-user-images.githubusercontent.com/1275028/350861197-58fcde11-472f-471b-abdb-a28824bb4ab7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwMTQ0MzIsIm5iZiI6MTcyNDAxNDEzMiwicGF0aCI6Ii8xMjc1MDI4LzM1MDg2MTE5Ny01OGZjZGUxMS00NzJmLTQ3MWItYWJkYi1hMjg4MjRiYjRhYjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MThUMjA0ODUyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTgyMGNlMGMwNzU0OWUyMGM4NmU1NTkwOWMwOGJhNDJkZmM3MTZhN2NlNzM0OWQxODcyYzZjMTMyNDUzMjVmZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.1xYlhXf1P0J4Xo7hDaPaNmr1V77k-KvwajScOAVxhJ8) The dialogue box only shows up once per Plasma session (which I think is fine).
> 
> This dialogue box is relatively new (maybe a month or so) auto-type would only work with X11 apps running in XWayland (I used to auto-type into Chromium) before I noticed this dialogue box. Hope this workaround helps anyone while the PR to get this properly implemented.

How exactly do you get this dialogue box to pop up? I have been trying to get autotype for wayland on plasma6 working for over a month.

If I start keepass with the -platform xcb argument as per the documentation then yes I can get the auto type tab however no apps, Wayland xwayland or otherwise actually work.

I never get any window asking for permission, When I press the global autotype shortcut in any window aside frim keepass itself, I don't see any errors it just does not work. If I try the Entry-Level Auto-Type, same result, I select the entry, keepassxc window minimizes then nothing previous window appears however it does not get any password.

I am currently on KDE Plasma 6.1.3. I would love to get this working it is perpahs the last major app that does not work with this environment.

**Comment by q-wertz** [comment]
For me the Auto-Type started working "magically" a few weeks ago.

I first get a pop-up from keepass, then one from the OS asking for permission and after confirming I can use autotype.
I'm on Wayland with gnome 46 (gdm 46.2-3)

Other maybe relevant packages:
- gnome-browser-connector 42.1-5
- gnome-desktop 1:44.0-1
- gnome-desktop-4 1:44.0-1
- gnome-desktop-common 1:44.0-1
- mutter 46.3.1-1.1
- xdg-desktop-portal-gnome 46.2-2

KeepassXC is installed via yay (not as flatpak) and keepassxc is started without any command line parameters/arguments (as far as I know)

**Comment by arigit** [comment]
> > When I run KeePassXC from the terminal using the command `keepassxc -platform xcb` I have get the auto type options enabled again (I know this is supposed to work). I am on KDE Plasma 6 running Wayland and am able to use the autotype function when in this mode to type into even what should be Wayland windows (Konsole, Kate, Kwrite, Firefox) after accepting a dialogue box which asks me to accept a remote control request. ![Screenshot_20240722_120525](https://private-user-images.githubusercontent.com/1275028/350861197-58fcde11-472f-471b-abdb-a28824bb4ab7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwMTQ0MzIsIm5iZiI6MTcyNDAxNDEzMiwicGF0aCI6Ii8xMjc1MDI4LzM1MDg2MTE5Ny01OGZjZGUxMS00NzJmLTQ3MWItYWJkYi1hMjg4MjRiYjRhYjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MThUMjA0ODUyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTgyMGNlMGMwNzU0OWUyMGM4NmU1NTkwOWMwOGJhNDJkZmM3MTZhN2NlNzM0OWQxODcyYzZjMTMyNDUzMjVmZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.1xYlhXf1P0J4Xo7hDaPaNmr1V77k-KvwajScOAVxhJ8) The dialogue box only shows up once per Plasma session (which I think is fine).
> > This dialogue box is relatively new (maybe a month or so) auto-type would only work with X11 apps running in XWayland (I used to auto-type into Chromium) before I noticed this dialogue box. Hope this workaround helps anyone while the PR to get this properly implemented.
> 
> How exactly do you get this dialogue box to pop up? I have been trying to get autotype for wayland on plasma6 working for over a month.
> 
> If I start keepass with the -platform xcb argument as per the documentation then yes I can get the auto type tab however no apps, Wayland xwayland or otherwise actually work.
> 
> I never get any window asking for permission, When I press the global autotype shortcut in any window aside frim keepass itself, I don't see any errors it just does not work. If I try the Entry-Level Auto-Type, same result, I select the entry, keepassxc window minimizes then nothing previous window appears however it does not get any password.
> 
> I am currently on KDE Plasma 6.1.3. I would love to get this working it is perpahs the last major app that does not work with this environment.

Same experience on Ubuntu 24.04.1, both firefox and keepassxc (2.7.9) installed as flatpaks. I added "QT_QPA_PLATFORM=xcb" to keepassxc enviroment via flatseal. The Autotype tab shows, but autotype is not working at all, from either global key or triggered from the menu context when right-clicking a DB entry

**Comment by ag-gaphp** [comment]
I'm on Arch, using KDE/Plasma

When installed via Discover/Flatpak, I can't get the auto-type tab to appear under any circumstance. Have tried setting the env var and adding the command flag to no avail.

When installed via pacman (kxc v2.7.9), the auto-type tab is already there by default, and when the database is locked it picks up the global hotkey I have set, asking to unlock the database. However, once unlocked, auto-type does nothing. I still tried adding `-platform xcd` to the command arguments, and setting `QT_QPA_PLATFORM=xcb`, but neither changes the behavior.

I get this log output when I start keepassxc (installed via pacman) without any arguments:
```
qt.qpa.plugin: Could not find the Qt platform plugin "wayland" in ""
```

**Comment by droidmonkey** [comment]
https://ardasevinc.hashnode.dev/launch-flatpak-apps-with-custom-args-and-environment-variables

**Comment by q-wertz** [comment]
> I'm on Arch, using KDE/Plasma

On my Manjaro/Gnome installation via yay AutoType is working and I get
```
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
```
when starting keepassxc

It seems to run in Xwayland mode and the error (that also @dasistwas has) seems to be unrelated.

EDIT:
BTW: When I use `keepassxc -platform xcb` AutoType works directly without the OS asking beforehand for permission.

**Comment by agowa** [comment]
> BTW: When I use keepassxc -platform xcb AutoType works directly without the OS asking beforehand for permission

I had to grant permission. That's probably because you already granted it...

**Comment by q-wertz** [comment]
> I had to grant permission. That's probably because you already granted it...

No... After closing the permission is lost and I have to regrant it when using it without `-xbc`.
Also there is a "sharing" indication icon when the permission is granted (like with microphone and camera)

**Comment by arigit** [comment]
> > > When I run KeePassXC from the terminal using the command `keepassxc -platform xcb` I have get the auto type options enabled again (I know this is supposed to work). I am on KDE Plasma 6 running Wayland and am able to use the autotype function when in this mode to type into even what should be Wayland windows (Konsole, Kate, Kwrite, Firefox) after accepting a dialogue box which asks me to accept a remote control request. ![Screenshot_20240722_120525](https://private-user-images.githubusercontent.com/1275028/350861197-58fcde11-472f-471b-abdb-a28824bb4ab7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwMTQ0MzIsIm5iZiI6MTcyNDAxNDEzMiwicGF0aCI6Ii8xMjc1MDI4LzM1MDg2MTE5Ny01OGZjZGUxMS00NzJmLTQ3MWItYWJkYi1hMjg4MjRiYjRhYjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MThUMjA0ODUyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTgyMGNlMGMwNzU0OWUyMGM4NmU1NTkwOWMwOGJhNDJkZmM3MTZhN2NlNzM0OWQxODcyYzZjMTMyNDUzMjVmZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.1xYlhXf1P0J4Xo7hDaPaNmr1V77k-KvwajScOAVxhJ8) The dialogue box only shows up once per Plasma session (which I think is fine).
> > > This dialogue box is relatively new (maybe a month or so) auto-type would only work with X11 apps running in XWayland (I used to auto-type into Chromium) before I noticed this dialogue box. Hope this workaround helps anyone while the PR to get this properly implemented.
> > 
> > 
> > How exactly do you get this dialogue box to pop up? I have been trying to get autotype for wayland on plasma6 working for over a month.
> > If I start keepass with the -platform xcb argument as per the documentation then yes I can get the auto type tab however no apps, Wayland xwayland or otherwise actually work.
> > I never get any window asking for permission, When I press the global autotype shortcut in any window aside frim keepass itself, I don't see any errors it just does not work. If I try the Entry-Level Auto-Type, same result, I select the entry, keepassxc window minimizes then nothing previous window appears however it does not get any password.
> > I am currently on KDE Plasma 6.1.3. I would love to get this working it is perpahs the last major app that does not work with this environment.
> 
> Same experience on Ubuntu 24.04.1, both firefox and keepassxc (2.7.9) installed as flatpaks. I added "QT_QPA_PLATFORM=xcb" to keepassxc enviroment via flatseal. The Autotype tab shows, but autotype is not working at all, from either global key or triggered from the menu context when right-clicking a DB entry


An update on this: there is an ugly workaround that can be used in wayland for keepassxc (flatpak) and firefox (flatpak) autotype to work: running both on xwayland, this is easy to achieve using Flatseal and removing the "Wayland Socket" permission from both Firefox and KeepassXC. Global autotype works well, as expected. Ubuntu 22.04 / nvidia 555

Of course, doing that is bad specially for firefox, it's back to the past (not as smooth (noticeably) browsing pages, security implications etc). So truly hopeful we will get native wayland autotype via a portal working sometime soon. In my environment could not manage yet to get the "prompt" that some folks report above.

**Comment by e-pirate** [comment]
Recently Gentoo marked Plasma 6 as Stable and masked KDE/Plasma 5 forcing users to update. I faced numerous problems with Wayland even it is Plasma version 6.1. Some were solved, some can be with a little bit of effort, some can be ignored or accepted. But non working AutoType function of KeepAssXC is an absolute blocker for me. Been a duty system administrator on I huge production, I'm forced to use AutoType in various combination primly for login into Windows machines via xfreerdp (we all hate this), typing in root passwords in IPMI sessions for rescuing stack servers and dozens of other scenarios when browser is no use and only AutoType is a solution. Having non-working AutoType simply making impossible to use Wayland.
So we have:
- wayland-1.23.0-r1
- keepassxc-2.7.9
- qtbase-6.7.2
- xcb-util-keysyms-0.4.1
- plasma-workspace-6.1.4-r1
If I start `keepassxc -platform xcb` I can now see the AutoType menu, but as many others when I select something there or just hit Ctrl-Shift-V, keepassxc window get minimized, but nothing typed into previous window and no 'Remote control request' window appear.
Any ideas what is wrong?

**Comment by tukanos** [comment]
> Recently Gentoo marked Plasma 6 as Stable and masked KDE/Plasma 5 forcing users to update. I faced numerous problems with Wayland even it is Plasma version 6.1. Some were solved, some can be with a little bit of effort, some can be ignored or accepted. But non working AutoType function of KeepAssXC is an absolute blocker for me. Been a duty system administrator on I huge production, I'm forced to use AutoType in various combination primly for login into Windows machines via xfreerdp (we all hate this), typing in root passwords in IPMI sessions for rescuing stack servers and dozens of other scenarios when browser is no use and only AutoType is a solution. Having non-working AutoType simply making impossible to use Wayland. So we have:
> 
>     * wayland-1.23.0-r1
> 
>     * keepassxc-2.7.9
> 
>     * qtbase-6.7.2
> 
>     * xcb-util-keysyms-0.4.1
> 
>     * plasma-workspace-6.1.4-r1
>       If I start `keepassxc -platform xcb` I can now see the AutoType menu, but as many others when I select something there or just hit Ctrl-Shift-V, keepassxc window get minimized, but nothing typed into previous window and no 'Remote control request' window appear.
>       Any ideas what is wrong?

I see the issue somewhere between the Wayland/X compatibility layer.

**Comment by kde-baskets-user** [comment]
May I kindly  ask someone to summarize in precise, technical terms the current status of the Auto-type feature on Wayland? 

- Does Auto-type  support pure Wayland mode?  Does it support both Plasma  6 and Gnome? If so, how was it implemented?
- Does the PR linked somewhere above address this issue and what's its fate?
- If it's not supported, please provide command line options to enable the Xwayland workaround  for both Gnome and KDE Plasma 6. 
- Do native, flatpack or snap installations make any difference in this regard?

You'd help enormously many users  by clarifying things.

**Comment by mareksamec** [comment]
> May I kindly ask someone to summarize in precise, technical terms the current status of the Auto-type feature on Wayland?
> 
> * Does Auto-type  support pure Wayland mode?  Does it support both Plasma  6 and Gnome? If so, how was it implemented?
> * Does the PR linked somewhere above address this issue and what's its fate?
> * If it's not supported, please provide command line options to enable the Xwayland workaround  for both Gnome and KDE Plasma 6.
> * Do native, flatpack or snap installations make any difference in this regard?
> 
> You'd help enormously many users by clarifying things.

I just tried KeePassXC 2.7.9 running on the Plasma 6 and Arch linux versions bellow and it works out of the box. What I needed to do is autostart my keepassxc with this argument `keepassxc -platform xcb` or start it with env variable like this `QT_QPA_PLATFORM=xcb keepassxc`. If your distro has recent enough version, it should work. Regarding flatpak or snaps. I'm not using those, but [this comment](https://github.com/keepassxreboot/keepassxc/issues/2281#issuecomment-2334006404) above seems to explain how to make it work on flatpak.

```
Operating System: Arch Linux 
KDE Plasma Version: 6.1.5
KDE Frameworks Version: 6.6.0
Qt Version: 6.7.2
Kernel Version: 6.10.10-arch1-1 (64-bit)
```

**Comment by e-pirate** [comment]
@mareksamec no, it does not work that way. I have all Plasma components equal or grater then 6.1.5, qtbase-6.7.2-r4, kde-frameworks 6.5.0 and running keepasxc with `-platform xcb' does not make autotype work, just enables the menu option back. Maybe frameworks are older then needed...

**Comment by tmccombs** [comment]
I think with the xcb platform it will work for other applications using Xwayland, but not native wayland applications.

**Comment by ag-gaphp** [comment]
> May I kindly ask someone to summarize in precise, technical terms the current status of the Auto-type feature on Wayland?
> 
> * Does Auto-type  support pure Wayland mode?  Does it support both Plasma  6 and Gnome? If so, how was it implemented? **A: It does not**
> 
> * Does the PR linked somewhere above address this issue and what's its fate? **A: [You can always get the latest info from the PR itself](https://github.com/keepassxreboot/keepassxc/pull/10905)**. TL;DR: it's not ready yet, as it is still in open status.
> 
> * If it's not supported, please provide command line options to enable the Xwayland workaround  for both Gnome and KDE Plasma 6. **A: There are none**
> 
> * Do native, flatpack or snap installations make any difference in this regard? **A: Makes no difference**
> 
> 
> You'd help enormously many users by clarifying things.

I get it's a long thread, but these are all pretty well clarified in the latest comments already. I went ahead and answered inline though.

If you need auto-type over anything else, use pure X11 instead of any wayland.

**Comment by odomingao** [comment]
Hi, anyone got it to work with hyprland?

**Comment by jakob1379** [comment]
Remember that enabling auto typing has security implications, not just on Wayland, but also under X11. Now that that has been said, I have found the simplest solution to be to update the "share/applications/org.keepassxc.KeePassXC.desktop" with 

```toml
Exec=env QT_QPA_PLATFORM=xcb keepassxc %f
```

This ensures if the application is started *correctly*. 

The issue by setting the `QT_QPA_PLATFORM=xcb` in you global env is that it will make all apps want to start up in something else that Wayland, which is not what you want.

**Comment by rusty-snake** [comment]
FWIW: `Exec=keepassxc -platform xcb` should do the same.

**Comment by droidmonkey** [comment]
Thank you for all the comments, we have a working solution to this in development. Please only comment if you have information that can support improving the working solution.

**Comment by DomingoMontoya** [comment]
I have an issue using it with wayland where it doesn't seem to be hitting the shift key? Any time there's a shift character, it types w/o it.

What I expect it to type `1!2@3#4$5%6^7&8*9(0)-_` vs what it actually types `11223344556677889(0)--`.

**Comment by lennybacon** [comment]
> Remember that enabling auto typing has security implications, not just on Wayland, but also under X11. Now that that has been said, I have found the simplest solution to be to update the "share/application/org.keepassxc.KeePassXC.desktop" with
> 
> Exec=env QT_QPA_PLATFORM=xcb keepassxc %f
> 
> This ensures if the application is started _correctly_.
> 
> The issue by setting the `QT_QPA_PLATFORM=xcb` in you global env is that it will make all apps want to start up in something else that Wayland, which is not what you want.

The path is missing a `s` - it is `applications` not `application`. The full path is: `~/.local/share/applications/org.keepassxc.KeePassXC.desktop` or `/usr/share/applications/org.keepassxc.KeePassXC.desktop`

**Comment by gongoscho** [comment]
Hello,

for your information. In the beginning on Debian 12 Stable with KDE Plasma (Wayland) Auto Type worked with the parameter “keepassxc -platform xcb”.

**After a kernel update not anymore!**

Unfortunately I don't know with which version it stopped working.

grEEtZ,
gongoscho

**Comment by mpeter50** [comment]
Thats most probably not related to the kernel update, as the kernel should have nothing to do with this.

But also, no, `-platform xcb` (and a few environment variables too) didnt make it work on Wayland either. The only thing that does is to request Keepass to start in X11 mode, so that it can autotype in the few windows that also run in X11 mode, but even then it wasnt able to work with any window on your system.
If it doesnt work now, thats probably because the programs you used with autotype switched to Wayland mode.

**Comment by lennybacon** [comment]
> Thank you for all the comments, we have a working solution to this in development. Please only comment if you have information that can support improving the working solution.

Just to clarify, what is the plan and strategy towards Wayland support? Just have it working out of the box for X-Windows? Or full support for Wayland?

**Comment by droidmonkey** [comment]
Full support for typing when running on Wayland. Unknown support for window targeting via window title. Unknown support for switching windows on users behalf. 

PR: https://github.com/keepassxreboot/keepassxc/pull/10905

**Comment by DomingoMontoya** [comment]
> Thats most probably not related to the kernel update, as the kernel should have nothing to do with this.
> 
> But also, no, `-platform xcb` (and a few environment variables too) didnt make it work on Wayland either. The only thing that does is to request Keepass to start in X11 mode, so that it can autotype in the few windows that also run in X11 mode, but even then it wasnt able to work with any window on your system. If it doesnt work now, thats probably because the programs you used with autotype switched to Wayland mode.

Although, you an set programs to run in xwindow mode with environment variables, for the things you need kepass to autotype to now. Just run `QT_QPA_PLATFORM=xcb GDK_BACKEND=x11 [program]` and it'll run in xwindows mode. Then, as long as you have keepassxc also running in xcb xwindows mode, it can autotype to those programs fine.

**Comment by agowa** [comment]
@DomingoMontoya is `GDK_BACKEND=x11` required? I've only `QT_QPA_PLATFORM=xcb` set via `/etc/environment` and everything appears to work just fine for everything.

**Comment by mpeter50** [comment]
`QT_QPA_PLATFORM` only affcets Qt based programs, like KDE programs. `GDK_BACKEND` is for GTK based programs, like Firefox, anything Gnome, EasyEffefcts. Java programs I think have yet another way, and then there are also some more other GUI SDKs.

**Comment by DomingoMontoya** [comment]
> [@DomingoMontoya](https://github.com/DomingoMontoya) is `GDK_BACKEND=x11` required? I've only `QT_QPA_PLATFORM=xcb` set via `/etc/environment` and everything appears to work just fine for everything.

As @mpeter50 said above, they're for different graphics tool sets. QT vs GTK.  When I'm in wayland I just set them on a per program basis where I know I'm going to need autotype. That way I don't get any ill intended effects where I can avoid them.

One thing that stinks, is that Firefox running in xwindows mode has pretty crummy scrolling vs wayland. But I live with it to have autotype work for keepass in firefox.

So, I basically just enable those for konsole (for ssh sessions), Firefox for MeshCentral, rustdesk, remmina, and let everything else run in wayland mode.

**Comment by mpeter50** [comment]
> One thing that stinks, is that Firefox running in xwindows mode has pretty crummy scrolling vs wayland. But I live with it to have autotype work for keepass in firefox.

Did you try the browser extension? So far it works for me, but I dont use websites where I would need to fill in complicated forms automatically.

**Comment by DomingoMontoya** [comment]
> > One thing that stinks, is that Firefox running in xwindows mode has pretty crummy scrolling vs wayland. But I live with it to have autotype work for keepass in firefox.
> 
> Did you try the browser extension? So far it works for me, but I dont use websites where I would need to fill in complicated forms automatically.

@mpeter50 , yes. I use it all the time. Thanks. It's just that I need the autotype in the browser, because I use MeshCentral to remote desktop and terminal into over 100 machines, so I like to be able to send the autotype remotely to remote lock screens, applications, etc.

**Comment by ccoenen** [comment]
and then there's software keys, wifi passwords, ssh credentials, desktop-software outside of the browser that requires logins...

Autotype has so many uses that can't be covered by the browser plugin alone. (Yes, I'm also using the browser plugin).

**Comment by mayonezo** [comment]
@ccoenen This is actually the feature that keeps me from migrating from keepass2/x to keepassxc/wayland.

**Comment by mpeter50** [comment]
If it works with Keepass2, shouldnt it also work with KeepassXC when it is started in X11 mode?

**Comment by mayonezo** [comment]
As far as I know #10905 is not merged, yet?

**Comment by mpeter50** [comment]
It is not, but since Keepass 2 runs in X11 mode (it does on my machine, and I think Mono apps dont support Wayland), and so it cannot type to Wayland windows, then it has the same autotype capabilities as KeepassXC when started in X11 mode.

**Comment by mayonezo** [comment]
That is why I still run an X11 session with keepass2 after finding out, that autotype does not work with wayland.

**Comment by sudhirkhanger** [comment]
😢 #10905 seems stale. Last comment is 3-4 months old.

**Comment by jidckii** [comment]
This ticket is already 7 years old — it seems we’ll never see this problem resolved.

**Comment by monotek** [comment]
I guess when x11 is gone in all distros...

**Comment by kde-baskets-user** [comment]
> I guess when x11 is gone in all distros...

It's a chicken-and-egg problem: AutoType is very basic functionality,  no way Wayland is "ready" w/o it, and no way X11 is ditched w/o any alternatives. 

All these bizarre non-reasons and deficiencies in architecture or development/governing processes related to Wayland, which make it hard to provide AutoType, just mean that there are many other UX problems with it anyway.

**Comment by michaelk83** [comment]
> 😢 [#10905](https://github.com/keepassxreboot/keepassxc/pull/10905) seems stale. Last comment is 3-4 months old.

It's waiting on some MRs on the KDE side, particularly [kwin!6992](https://invent.kde.org/plasma/kwin/-/merge_requests/6992). The devs on both sides are busy with other things.

**Comment by e-pirate** [comment]
While Wayland me a grate win and a nice future due to an architecture redesign it is a complete failure cause of the same reason. X11 had a huge well-working legacy that many useful things reside on. For example non working auto-type affects personally me in two different ways: non working keepassxc is the one (there is no way I can switch to Wayland because if this), the other sort of this problem is non-working xbindkeys: I have 12 button Logitech Revolution mouse (actually, two of them) heavily scripted to cover all my preferences on button pressings for many different applications. I've been using this set of xbindkeys + xprop + xvkbd for more then a decade and got used to my mouse behavior so much that I struggle every time I have to use different mouse even pressing non-existing buttons.

The other problem with Wayland is that there is no session save-restore functionality for some (all?) applications. For example every time I restart KDE/Plasma I get my konsole started with one empty tab while I have 8+ in special order and dedicated colors for half of them - this makes your life easier when you have to handle 20K server production of several semi-merged companies. I have to set up things that used to work out of a box for years and years before.

There many other annoying things that keeps people who need some not very popular / unexpected stuff to work for theirs professional activities. Lucky we are X still supported. I have to stick with X until auto-type and sessions restoring are fixed.

**Comment by jixiuf** [comment]
autotype work for me with my fork of 

![Image](https://github.com/user-attachments/assets/b4327a84-d515-4bb2-8c4e-6e2c7372b130)

https://github.com/jixiuf/kpmenu
[dotool](https://git.sr.ht/~jixiuf/dotool) and rofi 

```
XDG_CONFIG_HOME=~/.config kpmenu  --autotype 
# or  confirm with rofi 
XDG_CONFIG_HOME=~/.config kpmenu  --autotype --autotypeConfirm
```

my ~/.config/kpmenu/config
```
# You can store it into $HOME/.config/kpmenu/config
# Supported: dmenu, rofi, wofi, custom
menu = "custom"
# Supported: xsel, wl-clipboard, custom
clipboardTool = "wl-clipboard"
clipboardTimeout = "15s"
nocache = false
cacheOneTime = false
cacheTimeout = 60
nootp = false
daemon=true
customAutotypeWindowID=""" hyprctl activewindow -j| jq -c '"\\(.title) - \\(.class)"' """
customAutotypeTyper="dotoolc"
autotype=false

# Always prompt user to select the entry to autotype
# autotypeNoAuto=true
# Always confirm autotype, even when there's only 1 selection
# autotypeConfirm=true
# Executable of menus used to prompt actions
# customPromptPassword =""" sh -c "gpg -d ~/.password-store/bitwarden.com.gpg|head -n 1" """
customPromptPassword =""" echo -n '' """
# https://github.com/username/vmacs/blob/master/linux/config/rofi/config-bwmenu.rasi
customPromptMenu =  """rofi -dmenu  -i  -config ~/.config/rofi/config-bwmenu.rasi """
customPromptEntries = """rofi -dmenu  -i -mesg "C-7:user,C-8:passwd C-9:TOTP C-RET:passwd+RET C-0:URL" -p "Select entry" -config ~/.config/rofi/config-bwmenu.rasi -kb-custom-1 "Control+7" -kb-custom-2 "Control+8" -kb-custom-3 "Control+9" -kb-custom-4 "Control+Return" -kb-custom-5 "Control+0" """
customPromptFields  = """rofi -dmenu  -i -mesg "C-7:user,C-8:passwd C-9:TOTP C-RET:passwd+RET C-0:URL" -p "Select entry" -config ~/.config/rofi/config-bwmenu.rasi -kb-custom-1 "Control+7" -kb-custom-2 "Control+8" -kb-custom-3 "Control+9" -kb-custom-4 "Control+Return" -kb-custom-5 "Control+0" """
# Executable of clipboard commands
#customClipboardCopy = 
#customClipboardPaste = 
#customClipboardClean = 

passwordBackground = "black"
textPassword = "密码"
textMenu = "选择"
textEntry = "条目"
textField = "字段"
formatEntry = "{Title} - {UserName}"
#argsPassword =
#argsMenu =
#argsEntry =
#argsField =

#database =
keyFile ="/home/username/Documents/jianguo/jianguo/keepass/pass.key"
# password =""
# this need command "ykchalresp" from Yubikey Personalization
keyFileData="/home/username/Documents/jianguo/jianguo/keepass/pass.key"
# keyFileData=""" ykchalresp -x -2 -H %salt || echo -n "%salt"|xxd -r -p | openssl dgst -sha1 -hmac "$(gpg --pinentry-mode ask -d ~/Documents/jianguo/jianguo/keepass/keepassxc-cr-recovery/yubikey-otp-challenge-response-secret.gpg|xxd -r -p)"|awk '{print $2}' """
fieldOrder = "Password UserName URL"
fillOtherFields = true
#FillBlacklist =
database ="/home/username/Documents/jianguo/jianguo/keepass/pass.kdbx"
```

**Comment by tecbr** [comment]
Auto-Type works on Wayland! [As stated in the documentation](https://keepassxc.org/docs/KeePassXC_UserGuide#_auto_type):

> Auto-Type will be disabled when run with a Wayland compositor on Linux. To use Auto-Type in this environment, you must set QT_QPA_PLATFORM=xcb or start KeePassXC with the -platform xcb command-line flag.

But for me (Ubuntu 24.04) it only worked with QT_QPA_PLATFORM=xcb

Here's my code to fix the application shortcut:

`sudo sed -i "s|Exec=keepassxc %f|Exec=env QT_QPA_PLATFORM=xcb keepassxc -platform xcb %f|g" /usr/share/applications/org.keepassxc.KeePassXC.desktop`

**Comment by rusty-snake** [comment]
@tecbr that is exactly the definition of auto-type does not work with Wayland. QT_QPA_PLATFORM=xcb means your using XWayland.

**Comment by nbaak** [comment]
I am using Tuxedo OS which is based on Ubuntu 24.04 with wayland and the autotype does NOT work on my end. I wrote a script to set the Platform stuff but it does not work for me. The right window is selected (the color of the title bar changes) but no password is typed

```bash
#!/bin/bash

QT_QPA_PLATFROM=xcb

keepassxc -platform xcb
```

**Comment by TLATER** [comment]
You're only setting the variable in the current shell, it's not inherited by the child process. Try to `export` it, or prefix the command invocation with it. It also will only work on X11 windows because it uses X11 to do that.

Regardless, this is not a tech support forum, if you need help with bash there are plenty of other places to ask; please keep discussion on the actual topic, which is implementing support for auto-type when using the *native* Wayland backend.

Anyway, to summarize since this thread is once again going completely off-topic:

- This is not currently supported because of limitations of the Wayland protocols.
  - Wayland has a tighter threat model than X11. Allowing random applications to just interact with other application's windows is viewed as problematic because it makes writing keyloggers and such possible, so the platform doesn't currently support auto type.
  - This also means that complaining or asking about it here will not do anything, because the developers of this project literally cannot implement this; there is no way to make it work until the underlying problem is resolved.
- There is ongoing work in the Wayland ecosystem to solve this, at least for Plasma: https://invent.kde.org/plasma/kwin/-/merge_requests/6992
  - There is a matching PR that will use that work in this project: https://github.com/keepassxreboot/keepassxc/pull/10905
  - This stuff takes time, folks working on this are almost exclusively volunteers. Don't be surprised if it takes another year or two, especially since it will take at least that long for updates to distro packages to filter through. You'll have to be quite patient.
- If you desperately need this and can't wait, there are the following workarounds:
  - Setting `QT_QPA_PLATFROM=xcb` to use the Xwayland backend instead of the Wayland backend; this has significant limitations, such as only working on other Xwayland windows.
  - Browser plugins sidestep the issue, but obviously don't work on non-browser windows, or if the websites you visit don't have proper login forms.
  - You can try to implement hacks with auto-type workarounds that exist today - e.g. `/dev/uinput`, `dotool`, and similar tools (check e.g. the arch wiki for more) - as well as KeePassXC's secret service support yourself, but this is nontrivial and there's no easy generic solution that works across compositors and has no drawbacks (like requiring root permissions, or having security implications). If a good generic solution existed, this issue would likely already be closed.
  - KeePass apparently has some plugins which do the above. KeePassXC tries to be polished and secure, so those hacks aren't appropriate here, and hence the developers here are waiting for Wayland compositors to properly support this feature.

If you have nothing else to add, please resist the urge to reply, including thanking people, asking "are we there yet", or relieving your frustrations about wayland. There are plenty of other forums and social media platforms where you can do that without spamming people who don't care for that discussion, and are only subscribed here to know when things are actually progressing. Click the subscribe button if you want to do the same.

**Comment by hifi** [comment]
Anyone following this and waiting for the feature, please test #13359 if possible.
